V2.0 .dl file

emilkovacev opened the pull request:
Proposed changes
-------------------------------------------
Closes #91 
Adds a new .dl file structure

Types of changes
-------------------------------------------
What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

Checklist
-------------------------------------------
_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code._

- [x] My changeset covers only what is described above (no extraneous changes)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged downstream

Further comments
-------------------------------------------
I chose to add a new parser (according to the specification of the issue) because the original one was getting a bit cluttered. I chose to keep both of the structures, and modify existing implementations to match the current version of crpg (maintaining legacy versioning is one of the strengths of this project). I also added a bunch of improvements to the regex parsing, so that newlines and spaces are a lot more forgiving, which they should be.

The one improvement I would like to make, is have a generic parser in parser/__init__.py that grabs the version number and passes it on to the appropriate functions instead of having to call the parser manually. Maybe I'll do it later. *Maybe.*

zacharyLYH committed:
add_connection instead of add_oneway_connection

emilkovacev committed:
Revert "add_connection instead of add_oneway_connection"

WeirdAlex03 committed:
docs(PR template): Fix typo in instructions

The "x" to put in the boxes that apply under the Checklist heading
got lost at some point, so I added it back

WeirdAlex03 committed:
docs(PR template): Remove extreaneous "x"

I found the missing "x" from the checklist! It fell to the bottom
of the template, to the instructions for Further comments

Sploder12 committed:
updated .gitignore

Sploder12 committed:
Update .gitignore

now subdirectories are covered

Sploder12 committed:
Update .gitignore

All hidden files except github specific files

emilkovacev committed:
changed directory structure

emilkovacev committed:
changed directory structure

emilkovacev committed:
Added parser functionality

- Created new v2 parser
- Modified v1 parser
- Added v2 .dl example game
- modified builtin.py to include **kwargs implementation

Sploder12 commented on crpg/crpg.py:
```diff
@@ -0,0 +1,85 @@
+        node_a: Node, 
+        node_b: Node, 
+        connection_type_a: type = Connection, 
+        connection_type_b: type = Connection
```
shortened "conn_type" is used above, perhaps change these to "conn_type_a/b" or switch above to "connection_type" to keep consistency.

emilkovacev replied:
I changed conn -> connection

Sploder12 commented on crpg/setup.py:
```diff
@@ -1,13 +1,18 @@
     print(f.renderText('Classic RPG'))
 
-    while True:
+    while True: 
```
trailing whitespace

Sploder12 commented on crpg/parser/parser_v1.py:
```diff
@@ -0,0 +1,88 @@
+        if n_type == "starting":
+            starting = obj
+
+        node_mapping[n_alias] = obj
```
perhaps "n_id" or something similar would be better than "n_alias" here since it is a unique identifier?

emilkovacev replied:
switched n_alias -> node_id
And n_* -> node_*

Sploder12 commented on crpg/parser/parser_v2.py:
```diff
@@ -0,0 +1,131 @@
+        game.add_node(obj)
+
+        # generate the nodes
+        node_mapping[alias] = obj
```
same as parser_v1.py comment. "id" or something similar may be better than "alias" here for a unique key/identifier.

Sploder12 commented on crpg/parser/parser_v2.py:
```diff
@@ -0,0 +1,131 @@
+            game.current = obj
+
+        # add the node to the game
+        game.add_node(obj)
```
should use game.add_starter_node for the starting node

emilkovacev replied:
I'm deleting self.add_starter_node because of redundancy (I only ever need to call it once)

Sploder12 commented on crpg/parser/parser_v2.py:
```diff
@@ -0,0 +1,131 @@
+        node_a, n_conn, node_b = parse_connection(raw_conn)
+
+        node_a = node_mapping[node_a]
+        node_b = node_mapping[node_b]
```
conversion of node_a/b from str to Node may be confusing. Perhaps add variables like "node_a/b_id" to reduce confusion.

emilkovacev replied:
renamed variables appropriately

Sploder12 commented on games/game-3.py:
```diff
@@ -0,0 +1,7 @@
+import sys
+sys.path.insert(0, "../crpg")
```
This has mixed results, it doesn't work for me when I run using VSCode's default debugging behavior. Works fine running from command line in the directory. Using `os`'s path module with generated absolute paths may be more portable. 

perhaps using import with relative paths may work better, but that has it's own issues.

emilkovacev replied:
I think I may try to add pytest to this in the future, so that we can run **actual tests** in the future. I'll make that an issue

Sploder12 commented on crpg/parser/parser_v1.py:
```diff
@@ -0,0 +1,88 @@
+def generate(filename):
+    with open(filename, "r") as f:
+        contents = f.read()
+        nodes, connections = contents.split("\n+---\n+", maxsplit=1) 
```
not a regex split so splits on
```
+---
+
``` 
which is not intended

emilkovacev replied:
fixed regex declaration

Sploder12 commented on crpg/builtin.py:
```diff
@@ -12,48 +12,49 @@ def summary(self):
         self.from_node.connections.remove(self)
 
     def __str__(self):
-        return str(self.from_node) + " \-> " + str(self.to_node)
```
great catch!

 (I just realized I put the comment with the wrong side of the diff, sorry!)

Sploder12 reviewed (changes requested):
Really like these changes! I like where this project is headed.

Summary:
`crpg/parser/parser_v1.py` doesn't use regex split where it should.
`games/game-N.py` relative imports seem to be inconsistent.

various small naming suggestions

emilkovacev committed:
Semantic Changes

- Fixed variable names
- Removed trailing whitespaces
- Fixed regex definition
- Removed redundant method

emilkovacev committed:
merge v2.0 parser changes

emilkovacev committed:
Merge branch 'v2.0' of github.com:emilkovacev/Classic-RPG into v2.0

emilkovacev committed:
Merge branch 'v2.0' into v2.0-dl-file

Sploder12 commented on crpg/parser/parser_v1.py:
```diff
@@ -72,14 +71,13 @@ def parse_connections(chunks, node_mapping, game):
             c_func(game, node_a, node_b)
         else:
-            raise ConnectionError("Invalid .dl connection input")
+            raise ConnectionError("Invalid .dl connectionection input")
```
connectionection -> connection
(same thing on line 65 in same file)

Sploder12 commented on crpg/parser/parser_v2.py:
```diff
@@ -53,32 +58,32 @@ def parse_node(chunk):
 
     if connection is None or '' in connection.groups():
-        raise ConnectionError("Invalid .dl connection input")
+        raise ConnectionError("Invalid .dl connectionection input")
```
connectionection -> connection

Sploder12 commented on crpg/parser/parser_v2.py:
```diff
@@ -87,45 +92,45 @@ def node_setup(nodes, game):
 
-# setup game with nodes and connections
-def setup(nodes, conn, game):
+# setup game with nodes and connectionections
```
more connectionections! (and line 100 of same file)

Sploder12 commented on crpg/crpg.py:
```diff
@@ -1,31 +1,36 @@
         assert to_node in self.nodes
 
-        from_node.add_connection(to_node, connection_type)
+        from_node.add_connection(to_node, conn_type)
```
conn_type -> connection_type

Sploder12 reviewed (commented):
Looks good besides some connectionections!

emilkovacev committed:
fixed connectionections :)
